lib/api/objectPutPart.js

```diff
@@ -315,18 +345,17 @@ function objectPutPart(authInfo, request, streamingV4Params, log,
                 'content-md5': hexDigest,
                 'content-length': size,
             };
-            return metadata.putObjectMD(mpuBucketName, partKey, omVal, {}, log,
-                err => {
-                    if (err) {
-                        log.error('error putting object in mpu bucket', {
-                            error: err,
-                            method: 'objectPutPart::metadata.putObjectMD',
-                        });
-                        return next(err, destinationBucket);
-                    }
-                    return next(null, oldLocations, objectLocationConstraint,
-                        destinationBucket, hexDigest, prevObjectSize);
-                });
+            if (process.env.TEST_API_DATA === 'true') {
+                // Simulates a metadata failure for testing purposes.
+                memory.buckets.delete(mpuBucketName);
+            }
+            return safePutObjectMD(mpuBucketName, partKey, omVal, log, err => {
```

Comment:
minor: for clarity I would personally prefer keeping the code flow as-is and adding the rollback code directly in the callback of `putObjectMD`, because it's part of the global `objectPutPart` workflow rather than part of the `putObjectMD` operation. In other words it's hard to define what is a "safe" `putObjectMD` without knowing the context in which it is done (putting data first that might need a rollback).

Reply:
I agree and would like a generic method that takes a list of keys that need to be deleted. This can be re-used with the apis that are impacted Object PUT, Object PUT Copy Part, Object PUT Part, Object COPY.

Reply:
I can agree @jonathan-gramain that it is a bit better to be within the workflow, though it feels a lot more readable to have the steps broken down into smaller pieces. If adding directly to the workflow then, I suppose this means it won't be a common utility.

Reply:
The actual deletion function can be common to all backends (the one you created `deleteData` serves this purpose, though it might be augmented to fit other use cases regarding Rahul's comment), but it can be called directly from the workflow code in each place where it's needed, that's what I meant.